File: 97-interleaving-string/interleaving-string.java
class Solution {
  int m;
  int n;
  int l;
  boolean[][] dp;

  public boolean isInterleave(String s1, String s2, String s3) {
    m = s1.length();
    n = s2.length();
    l = s3.length();

    if(m + n != l) return false;

    dp = new boolean[m + 1][n + 1];

    dp[m][n] = true;


    for (int i = m; i >= 0; i--) {
      for (int j = n; j >= 0; j--) {

          if (i < m && s1.charAt(i) == s3.charAt(i + j)) {
            dp[i][j] = dp[i][j] || dp[i+1][j];
          }

          if (j < n && s2.charAt(j) == s3.charAt(i + j)) {
            dp[i][j] = dp[i][j] || dp[i][j+1];
          }

      }
    }

    return dp[0][0];
  }

  // private boolean dfs(int i, int j, int k, String s1, String s2, String s3) {

  //   if (i == m && j == n && k == l)
  //     return true;

  //   if (k >= l)
  //     return false;

  //   if (dp[i][j][k] != -1)
  //     return dp[i][j][k] == 1;

  //   dp[i][j][k] = res == true ? 1 : 0;

  //   return res;
  // }
}

File: 97-interleaving-string/interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        m = len(s1)
        n = len(s2)
        l = len(s3)

        if m + n != l:
            return False
        
        dp = [[0 for _ in range(n+1)] for _ in range(m+1)]
        dp[m][n] = True

        for i in range(m , -1 , -1):
            for j in range(n , -1 , -1):

                k = i + j
                if i < m and s1[i] == s3[k]:
                    dp[i][j] = dp[i][j] or dp[i+1][j]
                
                if j < n and s2[j] == s3[k]:
                    dp[i][j] = dp[i][j] or dp[i][j+1]

        return dp[0][0] 